### PR TITLE
Make: move compiler selection to the Makefile

### PR DESCRIPTION
**Pull request by tamird**
This change "demotes" the x86_64-unknown-linux-gnu toolchain, making
the builder's default behaviour more obvious. x86_64-unknown-linux-gnu
is now explicitly used to build releases, and since it is not binary-
compatible with the builder's default toolchain (current clang 3.8),
a new installsuffix is added to keep the artifacts separate.

### CODE REVIEW

**Review by benesch — APPROVED** [review]
After thinking about this more, I still think that the paths to `CC` and `CXX` should live somewhere other than the root Makefile, since that file shouldn't make any assumptions about the platform it's running on. But I think this is a step in the right direction, so LGTM.

I assume you tested the relevant bits of teamcity-release-upload.sh? The missing symbols in the race build make me a little nervous.

**Comment by benesch** [comment]
For posterity, by step in the right direction, I mean that with this change the release compiler paths now live in one place (the root Makefile), instead the Makefile (musl) _and_ the Dockerfile (non-musl).

File: Makefile [hunk 1]
```diff
@@ -63,13 +58,19 @@ LINKFLAGS ?=
...
+export CXX = /x-tools/x86_64-unknown-linux-gnu/bin/x86_64-unknown-linux-gnu-g++
+override LINKFLAGS += -s -w -extldflags '\''-static-libgcc -static-libstdc++'\'' -X github.com/cockroachdb/cockroach/pkg/build.typ=release
+override GOFLAGS += -installsuffix release
 else ifeq ($(TYPE),musl)
 # This tag disables jemalloc profiling. See https://github.com/jemalloc/jemalloc/issues/585.
 override TAGS += musl
 export CC  = /x-tools/x86_64-unknown-linux-musl/bin/x86_64-unknown-linux-musl-gcc
 export CXX = /x-tools/x86_64-unknown-linux-musl/bin/x86_64-unknown-linux-musl-g++
-override LINKFLAGS += -extldflags -static -X github.com/cockroachdb/cockroach/pkg/build.typ=release-musl
+override LINKFLAGS += -s -w -extldflags '\''-static'\'' -X github.com/cockroachdb/cockroach/pkg/build.typ=release-musl
```
Comment (tamird) [inline Makefile]: Lots of precedent here for this kind of quote literal, and double quotes have different meaning wrt expansion.

Comment (tamird) [inline Makefile]: Had to use single quotes to get through the many layers of shell expansion.

**Comment by tamird** [comment]
This is green, but merging it is likely to disrupt PRs with an older parent (because of cache pollution). Any ideas?

**Comment by bdarnell** [comment]
Reviewed 2 of 3 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[build/Dockerfile, line 103 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14945#-KhtugKrKyh1g9yrRzc9:-KhtugKrKyh1g9yrRzcA:b-8fe3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f62e136289cf264b61c44c9717356921e5c764/build/Dockerfile#L103)):*
> ```Dockerfile
>  && mkdir libcxx_msan && (cd libcxx_msan && cmake ../llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_USE_SANITIZER=Memory && make cxx -j$(nproc))
> 
> # Go assumes "gcc" and "g++", so we must teach it otherwise.
> ```

Why? We generally prefer to use the "native" compiler for the platform - gcc on linux and clang on macos/bsd. Should also comment about the relationship between the selection here and in the Makefile.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14945)*

**Comment by tamird** [comment]
Well we need clang to support msan; by the end of building, the image
didn't have gcc.

**Comment by tamird** [comment]
Updated to change all the TC targets to use release build settings. PTAL.

File: build/teamcity-jepsen.sh [hunk 2]
```diff
@@ -69,19 +69,20 @@ for test in "${tests[@]}"; do
...
         # If this begins to time out frequently, let's do this via nohup and poll.
-        ssh -o "ServerAliveInterval=60" -o "StrictHostKeyChecking no" -i "$HOME/.ssh/${KEY_NAME}" "ubuntu@${controller}" "${testcmd}"
-        if [ $? -ne 0 ]; then
+        #
+        # shellcheck disable=SC2029
+        if ssh -o "ServerAliveInterval=60" -o "StrictHostKeyChecking no" -i "$HOME/.ssh/${KEY_NAME}" "ubuntu@${controller}" "${testcmd}"; then
+            # Test passed. grab just the results file.
+            echo "Test passed. Grabbing minimal logs..."
+            scp -o "StrictHostKeyChecking no" -ri "$HOME/.ssh/${KEY_NAME}" "ubuntu@${controller}:jepsen/cockroachdb/store/latest/{test.fressian,results.edn,latency-quantiles.png,latency-raw.png,rate.png}" "${artifacts_dir}"
+        else
```
Comment (benesch) [inline build/teamcity-jepsen.sh]: Much nicer. 👍

File: build/builder.sh [hunk 3]
```diff
@@ -122,13 +122,23 @@ fi
...
 vols="${vols} --volume=${gocache}/pkg/docker_amd64:/usr/local/go/pkg/linux_amd64"
-vols="${vols} --volume=${gocache}/pkg/docker_amd64_msan:/usr/local/go/pkg/linux_amd64_msan"
 vols="${vols} --volume=${gocache}/pkg/docker_amd64_musl:/usr/local/go/pkg/linux_amd64_musl"
+vols="${vols} --volume=${gocache}/pkg/docker_amd64_release:/usr/local/go/pkg/linux_amd64_release"
+vols="${vols} --volume=${gocache}/pkg/docker_amd64_msan:/usr/local/go/pkg/linux_amd64_msan"
+vols="${vols} --volume=${gocache}/pkg/docker_amd64_musl_msan:/usr/local/go/pkg/linux_amd64_musl_msan"
+vols="${vols} --volume=${gocache}/pkg/docker_amd64_release_msan:/usr/local/go/pkg/linux_amd64_release_msan"
 vols="${vols} --volume=${gocache}/pkg/docker_amd64_race:/usr/local/go/pkg/linux_amd64_race"
+vols="${vols} --volume=${gocache}/pkg/docker_amd64_musl_race:/usr/local/go/pkg/linux_amd64_musl_race"
+vols="${vols} --volume=${gocache}/pkg/docker_amd64_release_race:/usr/local/go/pkg/linux_amd64_release_race"
```
Comment (benesch) [inline build/builder.sh]: It's probably time to do this in a loop.

**Review by benesch — APPROVED** [review]
LGTM!

File: build/builder.sh (hunk 3 above)
Comment (tamird) [inline build/builder.sh]: Yuge point. Done.

**Comment by benesch** [comment]
Last three commits LGTM too!

**Comment by jordanlewis** [comment]
I'm not a huge fan of making `TYPE=release` required everywhere. It makes the build process more confusing. Can we make `TYPE=release` the default? In what circumstances would we prefer building with the "default toolchain"? IMO we should make users who prefer the "default toolchain" specify it explicitly because it seems to be a very uncommon case.

---

Reviewed 1 of 3 files at r1, 10 of 12 files at r3, 3 of 3 files at r4, 3 of 3 files at r5, 1 of 1 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[Makefile, line 190 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14945#-KhwvIB6r2isGkQP1onM:-KhwvIB6r2isGkQP1onN:b-gxlb7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe8b142a370b2bd5482481e365309c06d869ed/Makefile#L190)):*
> ```Makefile
> .PHONY: dupl
> dupl: $(BOOTSTRAP_TARGET)
> 	$(FIND_RELEVANT) \
> ```

I've never noticed this target, but should it live in `make check`?

---

*[build/common.mk, line 54 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14945#-KhwvFWm1B5OKKc_BJRF:-KhwvFWnlGAeb8s3yDVt:b-hk98dr) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe8b142a370b2bd5482481e365309c06d869ed/build/common.mk#L54)):*
> ```Makefile
> # Docker for Mac, so we filter out the 20k+ UI dependencies that are
> # guaranteed to be irrelevant to save nearly 10s on every Make invocation.
> FIND_RELEVANT := find pkg -name node_modules -prune -o
> ```

Why does this need to be in `common.mk`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14945#-:-KhwwcrN7vBiYW9dtdO0:b52sido)*

**Comment by benesch** [comment]
We can't make `TYPE=release` the default, because `TYPE=release` hardcodes CC to a path that only exists in the builder image. We prefer building with the default toolchain when building locally, since there's no reason to build with an old libc unless building a release binary for wide distribution.

I think these TC scripts should have some centralized way of specifying `CC=/the/desired/build/cc` without polluting either the Dockerfile or the Makefile.

**Comment by jordanlewis** [comment]
Ah, I see.

Okay, I guess I rescind my objection. I am still a bit wary - forcing all the TC scripts to specify a special flag every time they run `make` (whether that's `TYPE` or `CC`) still feels a bit anti-patternish to me but I can't immediately see a compelling alternative.

And the `TYPE=release` flag definitely needs to be documented somewhere like `CONTRIBUTING.md` or some more suitable location.

---

Review status: 16 of 17 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[Makefile, line 63 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14945#-KhwzOTtpBjUL4OrspZV:-KhwzOTtpBjUL4OrspZW:b-w518y3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f51d376923ef20a02f642fdd7e2cc062cbfd7e42/Makefile#L63)):*
> ```Makefile
> # We use a custom toolchain to target old Linux and glibc versions. However,
> # this toolchain's libstdc++ version is quite recent and must be statically
> # linked to avoid depending on the target's available libstdc++.
> ```

I think it's important that you explain where these paths come from. This target is coupled to the builder image and as long as that's true I think we need to call it out here to prevent confusion when people try to make a release build on their own.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14945#-:-Khwy0AITLL1ngZh0miO:b6egunj)*

**Comment by tamird** [comment]
`TYPE=release` being required everywhere is just a mitigation strategy for the potential cache pollution issue which is specific to our TC environment.

Why would we document TYPE=release in contributing.md?

---

Review status: 14 of 17 files reviewed at latest revision, 7 unresolved discussions.

---

*[Makefile, line 190 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14945#-KhwvIB6r2isGkQP1onM:-KhwyzUakSxgvbw_BZm2:b2yrlzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe8b142a370b2bd5482481e365309c06d869ed/Makefile#L190)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I've never noticed this target, but should it live in `make check`?
</blockquote></details>

I don't think so. It's here as a convenience, not something that's enforced.

---

*[Makefile, line 63 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14945#-KhwzOTtpBjUL4OrspZV:-Khx1dVuqfttac4Y2Ng5:b37b60e) ([raw file](https://github.com/cockroachdb/cockroach/blob/f51d376923ef20a02f642fdd7e2cc062cbfd7e42/Makefile#L63)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think it's important that you explain where these paths come from. This target is coupled to the builder image and as long as that's true I think we need to call it out here to prevent confusion when people try to make a release build on their own.
</blockquote></details>

This is explained in the block comment on TYPE.

---

*[build/common.mk, line 54 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14945#-KhwvFWm1B5OKKc_BJRF:-Khwz3qKGE794GPa9-pl:bj4b1yf) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe8b142a370b2bd5482481e365309c06d869ed/build/common.mk#L54)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Why does this need to be in `common.mk`?
</blockquote></details>

I thought it would be used in protobuf.mk. Moved to Makefile.

---

*[build/Dockerfile, line 103 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14945#-KhtugKrKyh1g9yrRzc9:-KhwzZAvpIGAS46sY2zb:b-ukimfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f62e136289cf264b61c44c9717356921e5c764/build/Dockerfile#L103)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why? We generally prefer to use the "native" compiler for the platform - gcc on linux and clang on macos/bsd. Should also comment about the relationship between the selection here and in the Makefile.
</blockquote></details>

We need clang to support msan, and there's no reason to have both (I think). By the end of building, the image doesn't have gcc.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14945#-:-KhwzMCbzvygdqisgnSf:b-7icu1m)*

**Pull request merged**